hadoop-hdds/container-service/src/test/java/org/apache/hadoop/hdds/datanode/metadata/TestDatanodeCRLStoreImpl.java

```diff
@@ -0,0 +1,126 @@
+import org.bouncycastle.cert.X509v2CRLBuilder;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.File;
+import java.security.KeyPair;
+import java.security.cert.X509Certificate;
+import java.util.Date;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+/**
+ * Test class for {@link DatanodeCRLStoreImpl}.
+ */
+public class TestDatanodeCRLStoreImpl {
+  private File testDir;
+  private OzoneConfiguration conf;
+  private DatanodeCRLStore dnCRLStore;
+  private KeyPair keyPair;
+  private CRLApprover crlApprover;
+  private SecurityConfig securityConfig;
+
+  @Before
+  public void setUp() throws Exception {
+    testDir = GenericTestUtils.getRandomizedTestDir();
+    conf = new OzoneConfiguration();
+    conf.set(HddsConfigKeys.OZONE_METADATA_DIRS, testDir.getPath());
+    dnCRLStore = new DatanodeCRLStoreImpl(conf);
+    keyPair = KeyStoreTestUtil.generateKeyPair("RSA");
+    securityConfig = new SecurityConfig(conf);
+  }
+
+  @Before
+  public void initCRLApprover() {
+    crlApprover = new DefaultCRLApprover(securityConfig,
+        keyPair.getPrivate());
+  }
+
+  @After
+  public void tearDown() {
+    FileUtil.fullyDelete(testDir);
+  }
+
+  @After
+  public void destroyDbStore() throws Exception {
+    if (dnCRLStore.getStore() != null) {
+      dnCRLStore.getStore().close();
+    }
+  }
+  @Test
+  public void testCRLStore() throws Exception {
+    assertNotNull(dnCRLStore.getStore());
+
+    dnCRLStore.getCRLSequenceIdTable().put(OzoneConsts.CRL_SEQUENCE_ID_KEY, 5L);
+    Date now = new Date();
+    X509Certificate x509Certificate = generateX509Cert();
+    X509CertificateHolder caCertificateHolder =
+        new X509CertificateHolder(generateX509Cert().getEncoded());
+    X509v2CRLBuilder crlBuilder = new X509v2CRLBuilder(
+        caCertificateHolder.getIssuer(), now);
+    crlBuilder.addCRLEntry(x509Certificate.getSerialNumber(), now,
+        CRLReason.lookup(CRLReason.PRIVILEGE_WITHDRAWN).getValue().intValue());
+    dnCRLStore.getPendingCRLsTable().put(1L,
+        new CRLInfo.Builder()
+            .setCrlSequenceID(1L)
+            .setCreationTimestamp(now.getTime())
+            .setX509CRL(crlApprover.sign(crlBuilder))
+            .build());
+
+    assertEquals(5L, (long) dnCRLStore.getLatestCRLSequenceID());
+    assertEquals(1L, dnCRLStore.getPendingCRLs().size());
+    CRLInfo crlInfo = dnCRLStore.getPendingCRLs().get(0);
+    assertEquals(1L, crlInfo.getCrlSequenceID());
+    assertEquals(x509Certificate.getSerialNumber(),
+        crlInfo.getX509CRL().getRevokedCertificates()
+            .iterator().next().getSerialNumber());
+
+    // Test that restarting the store does not affect the data already persisted
+    dnCRLStore.stop();
+    dnCRLStore = new DatanodeCRLStoreImpl(conf);
+    assertEquals(5L, (long) dnCRLStore.getLatestCRLSequenceID());
+    assertEquals(1L, dnCRLStore.getPendingCRLs().size());
```

Comment:
Can we add a dnCRLStore.stop() at the end?

Reply:
Sure. I have added this line in the latest patch. But, it will be taken care of by the destroyDbStore() regardless after each test.